independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/PathTree.java

```diff
@@ -0,0 +1,71 @@
+package io.quarkus.paths;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.Collection;
+import java.util.function.Function;
+import java.util.jar.Manifest;
+
+public interface PathTree {
+
+    static PathTree of(Path p) {
+        try {
+            BasicFileAttributes fileAttributes = Files.readAttributes(p, BasicFileAttributes.class);
+            if (fileAttributes.isDirectory()) {
+                return new DirectoryPathTree(p);
+            }
+
+            return new FilePathTree(p);
+        } catch (IOException e) {
+            throw new IllegalArgumentException(p + " does not exist");
```

Comment:
This is probably the most likely case, but javadoc of says: `if an I/O error occurs`, so better include the IOE as the cause?

Reply:
Agreed